### PR TITLE
Adding Event Notifier base component

### PR DESCRIPTION
**Pull request by CrowleyRajapakse**
## Methodology

Adding event notifier interfaces which can be used to send various events data such as creation,updation,deletion of Subscriptions, Applications, APIs, Policies, Keys to Pilot node.

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java [hunk 1]
```diff
@@ -3188,6 +3195,12 @@ public void removeSubscription(SubscribedAPI subscription) throws APIManagementE
                         + ") removed from app " + appName;
                 log.debug(logMessage);
             }
+
+            SubscriptionEvent subscriptionEvent = new SubscriptionEvent(UUID.randomUUID().toString(),
```
Comment (tharindu1st) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java:3199]: These Events need to send after subscription approved

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java [hunk 2]
```diff
@@ -3408,6 +3421,12 @@ public int addApplication(Application application, String userId)
             Thread recommendationThread = new Thread(extractor);
             recommendationThread.start();
         }
+
+        ApplicationEvent applicationEvent = new ApplicationEvent(UUID.randomUUID().toString(), System.currentTimeMillis()
+                , APIConstants.EventType.APPLICATION_CREATE.name(), tenantId, applicationId, application.getName(), application.getTokenType(),
+                application.getTier());
```
Comment (tharindu1st) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java:3427]: same goes here as well

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java [hunk 3]
```diff
@@ -85,6 +85,7 @@
 import org.wso2.carbon.apimgt.impl.dto.TierPermissionDTO;
 import org.wso2.carbon.apimgt.impl.dto.WorkflowDTO;
 import org.wso2.carbon.apimgt.impl.factory.KeyManagerHolder;
+import org.wso2.carbon.apimgt.impl.notifier.events.*;
```
Comment (isharac) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java:88]: change * to exact class names

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java [hunk 4]
```diff
@@ -3906,6 +3936,11 @@ else if (APIConstants.API_KEY_TYPE_SANDBOX.equals(tokenType)) { // if
             APIUtil.logAuditMessage(APIConstants.AuditLogConstants.APPLICATION, appLogObject.toString(),
                     APIConstants.AuditLogConstants.UPDATED, this.username);
 
+            ApplicationRegistrationEvent applicationRegistrationEvent = new ApplicationRegistrationEvent(
```
Comment (isharac) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java:3939]: this event has to go to completeApplicationRegistration

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java [hunk 5]
```diff
@@ -4050,6 +4085,12 @@ else if (APIConstants.API_KEY_TYPE_SANDBOX.equals(tokenType)) {
             appLogObject.put("Generated keys for application", application.getName());
             APIUtil.logAuditMessage(APIConstants.AuditLogConstants.APPLICATION, appLogObject.toString(),
                     APIConstants.AuditLogConstants.UPDATED, this.username);
+
```
Comment (isharac) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java:4088]: this as well

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/internal/APIManagerComponent.java [hunk 6]
```diff
@@ -49,6 +49,8 @@
 import org.wso2.carbon.apimgt.impl.factory.KeyManagerHolder;
 import org.wso2.carbon.apimgt.impl.factory.SQLConstantManagerFactory;
 import org.wso2.carbon.apimgt.impl.handlers.UserPostSelfRegistrationHandler;
+import org.wso2.carbon.apimgt.impl.notifier.*;
```
Comment (isharac) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/internal/APIManagerComponent.java:52]: change to absolute imports

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/internal/APIManagerComponent.java [hunk 7]
```diff
@@ -169,6 +172,19 @@ protected void activate(ComponentContext componentContext) throws Exception {
             bundleContext.registerService(Axis2ConfigurationContextObserver.class.getName(), tenantLoadMessageSender, null);
             KeyMgtConfigDeployer keyMgtConfigDeployer = new KeyMgtConfigDeployer();
             bundleContext.registerService(Axis2ConfigurationContextObserver.class.getName(), keyMgtConfigDeployer, null);
+
+            //Registering Notifiers
+            Notifier subscriptionsNotifier = new SubscriptionsNotifier();
```
Comment (isharac) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/internal/APIManagerComponent.java:177]: we can merge these two lines
bundleContext.registerService(Notifier.class.getName(), new SubscriptionsNotifier(), null);

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/internal/APIManagerComponent.java [hunk 8]
```diff
@@ -731,5 +747,25 @@ private void setupAccessTokenGenerator(){
             ServiceReferenceHolder.getInstance().setAccessTokenGenerator(accessTokenGenerator);
         }
     }
+
+    @Reference(
+            name = "notifier.component",
+            service = Notifier.class,
+            cardinality = ReferenceCardinality.MULTIPLE,
+            policy = ReferencePolicy.DYNAMIC,
+            unbind = "removeNotifier")
```
Comment (isharac) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/internal/APIManagerComponent.java:756]: shouldn't this be removeNotifiers?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/notifier/events/APIEvent.java [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+
+    private String apiName;
+    private int apiId;
+    private String apiVersion;
+    private String apiContext;
+    private String apiProvider;
+    private String apiType;
+    private String apiStatus;
+
+    public APIEvent(String eventId, long timestamp, String type, int tenantId, String apiName, int apiId,
```
Comment (isharac) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/notifier/events/APIEvent.java:15]: Shall we change this to builder pattern?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/notifier/events/ApplicationEvent.java [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+    public ApplicationEvent(String eventId, long timestamp, String type, int tenantId,int applicationId,
+                            String applicationName, String tokenType, String applicationPolicy) {
+        this.eventId = eventId;
+        this.timeStamp = timestamp;
+        this.type = type;
+        this.tenantId = tenantId;
+        this.applicationId = applicationId;
+        this.applicationName = applicationName;
+        this.tokenType = tokenType;
+        this.applicationPolicy = applicationPolicy;
```
Comment (isharac) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/notifier/events/ApplicationEvent.java:20]: We need the group_id as well.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/APIUtil.java [hunk 11]
```diff
@@ -539,6 +535,23 @@ public static void updateAPIProductDependencies(API api, Registry registry) thro
...
+     * @param event        Event object
+     * @param notifierType eventType
+     */
+    public static void sendNotification(org.wso2.carbon.apimgt.impl.notifier.events.Event event, String notifierType) {
+        List<Notifier> notifierList = ServiceReferenceHolder.getInstance().getNotifiersMap().get(notifierType);
+        notifierList.forEach((notifier) -> {
+            try {
+                notifier.publishEvent(event);
+            } catch (NotifierException e) {
+                log.error("Error when publish " + event + " through notifier:" + notifierType + " .Error:" + e);
```
Comment (isharac) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/APIUtil.java:550]: need the space after the .

Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/APIUtil.java]: resolved

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/notifier/events/ApplicationEvent.java (hunk 10 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/notifier/events/ApplicationEvent.java:21]: resolved

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/internal/APIManagerComponent.java (hunk 8 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/internal/APIManagerComponent.java]: resolved

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/internal/APIManagerComponent.java (hunk 7 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/internal/APIManagerComponent.java]: resolved

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/internal/APIManagerComponent.java (hunk 6 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/internal/APIManagerComponent.java]: resolved

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java (hunk 4 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java:4098]: resolved

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java (hunk 5 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java]: resolved

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java (hunk 3 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java]: resolved

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java (hunk 1 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java:3362]: resolved

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java (hunk 2 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java]: resolved

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java (hunk 1 above)
Comment (tharindu1st) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java:3362]: still not fixed

**Comment by Arshardh** [comment]
Looks good, but try to add class comments if possible for newly created classes.

**Pull request merged**